src/stdlib/io.rs

```diff
@@ -69,4 +88,831 @@ pub fn load_io<'gc>(ctx: Context<'gc>) {
+                Err(err) => {
+                    stack.replace(
+                        ctx,
+                        (Value::Nil, err.to_string(), err.raw_os_error().unwrap_or(0)),
+                    );
+                    Ok(CallbackReturn::Return)
+                }
+            }
+        }),
+    );
+
+    io.set_field(
+        ctx,
+        "input",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            #[derive(Collect)]
+            #[collect(no_drop)]
+            struct InputOpenThen;
+
+            impl<'gc> Sequence<'gc> for InputOpenThen {
+                fn poll(
+                    self: Pin<&mut Self>,
+                    ctx: Context<'gc>,
+                    _exec: Execution<'gc, '_>,
+                    stack: Stack<'gc, '_>,
+                ) -> Result<SequencePoll<'gc>, Error<'gc>> {
+                    let file: Value = stack.get(0);
+
+                    if file.is_nil() {
+                        return Ok(SequencePoll::Return);
+                    }
+
+                    match file {
+                        Value::UserData(file) => {
+                            if let Ok(file) = file.downcast_static::<IoFile>() {
+                                IO_STATE.with(|state| state.replace_input(file.clone()));
+                                Ok(SequencePoll::Return)
+                            } else {
+                                Err("bad argument #1 to 'input' (file expected)"
+                                    .into_value(ctx)
+                                    .into())
+                            }
+                        }
+                        _ => Err("bad argument #1 to 'input' (file expected)"
+                            .into_value(ctx)
+                            .into()),
+                    }
+                }
+            }
+
+            if stack.is_empty() {
+                stack.replace(
+                    ctx,
+                    UserData::new_static(&ctx, IO_STATE.with(|state| state.input())),
+                );
+                return Ok(CallbackReturn::Return);
+            }
+
+            let value = stack.consume::<Value>(ctx)?;
+
+            match value {
+                Value::String(filename) => {
+                    let io: Table = ctx.globals().get(ctx, "io")?;
+                    let open: Callback = io.get(ctx, "open")?;
+
+                    stack.replace(ctx, (filename, "r"));
+                    Ok(CallbackReturn::Call {
+                        function: open.into(),
+                        then: Some(BoxSequence::new(&ctx, InputOpenThen)),
+                    })
+                }
+                Value::UserData(file) => {
+                    if let Ok(file) = file.downcast_static::<IoFile>() {
+                        IO_STATE.with(|state| state.replace_input(file.clone()));
+
+                        stack.replace(ctx, UserData::new_static(&ctx, file.clone()));
+                        Ok(CallbackReturn::Return)
```

Comment:
This could return the original UserData, rather than creating a new userdata for it.